### PR TITLE
[SYCL] Fix get_pointer_device for cases with descendent devices

### PR DESCRIPTION
**Pull request by sergey-semenov**
Looking through context members alone when searching for a specific device isn't enough anymore since now descendent devices of context members can be used within that context as well. Change the logic to look for the device in the cache instead.

### CODE REVIEW

**Comment by sergey-semenov** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1232

**Comment by sergey-semenov** [comment]
@cperkinsintel Could you please review this one?

**Comment by keryell** [comment]
What do you mean by "descendent devices"?
Subdevices?

**Comment by sergey-semenov** [comment]
> What do you mean by "descendent devices"? Subdevices?

Subdevices, subsubdevices, etc. It's a term used in the spec: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#descendent-device

**Comment by sergey-semenov** [comment]
@cperkinsintel @intel/llvm-reviewers-runtime Gentle ping

File: sycl/source/detail/platform_impl.cpp [hunk 1]
```diff
@@ -334,6 +336,17 @@ bool platform_impl::has(aspect Aspect) const {
   return true;
 }
 
+std::shared_ptr<device_impl> platform_impl::getDeviceImplHelper(
+    RT::PiDevice PiDevice, const std::shared_ptr<platform_impl> &PlatformImpl) {
+  for (const std::weak_ptr<device_impl> &DeviceWP : MDeviceCache) {
```
Comment (romanovvlad) [inline sycl/source/detail/platform_impl.cpp:341]: A comment saying that this function is expected to be called with `MDeviceMapMutex` locked would be nice.

**Review by romanovvlad — APPROVED** [review]
LGTM

File: sycl/source/detail/platform_impl.cpp (hunk 1 above)
Comment (sergey-semenov) [reply sycl/source/detail/platform_impl.cpp:341]: Agreed, I'll add it in a separate PR.

**Comment by sergey-semenov** [comment]
@intel/llvm-gatekeepers Could you please merge this?

**Pull request merged**